src/analysis/ExpressionExplorer.jl

```diff
@@ -344,419 +337,528 @@ Returns whether or not an assignment Expr(:(=),...) is assigning to a new functi
-        end
+    if is_function_assignment(ex)
+        # f(x, y) = x + y
+        # Rewrite to:
+        # function f(x, y) x + y end
+        return explore!(Expr(:function, ex.args...), scopestate)
+    end
 
-        val = ex.args[2]
-        # Handle generic types assignments A{B} = C{B, Int}
-        if ex.args[1] isa Expr && ex.args[1].head == :curly
-            assignees, symstate = explore_funcdef!(ex.args[1], scopestate)
-            innersymstate = union!(symstate, explore!(val, scopestate))
-        else
-            assignees = get_assignees(ex.args[1])
-            symstate = innersymstate = explore!(val, scopestate)
+    val = ex.args[2]
+    # Handle generic types assignments A{B} = C{B, Int}
+    if ex.args[1] isa Expr && ex.args[1].head::Symbol == :curly
+        assignees, symstate = explore_funcdef!(ex.args[1], scopestate)::Tuple{Vector{Symbol}, SymbolsState}
+        innersymstate = union!(symstate, explore!(val, scopestate))
+    else
+        assignees = get_assignees(ex.args[1])
+        symstate = innersymstate = explore!(val, scopestate)
+    end
+
+    global_assignees = get_global_assignees(assignees, scopestate)
+
+    # If we are _not_ assigning a global variable, then this symbol hides any global definition with that name
+    push!(scopestate.hiddenglobals, setdiff(assignees, global_assignees)...)
+    assigneesymstate = explore!(ex.args[1], scopestate)
+
+    push!(scopestate.hiddenglobals, global_assignees...)
+    push!(symstate.assignments, global_assignees...)
+    push!(symstate.references, setdiff(assigneesymstate.references, global_assignees)...)
+    filter!(!all_underscores, symstate.references)  # Never record _ as a reference
+
+    return symstate
+end
+
+function explore_modifiers!(ex::Expr, scopestate::ScopeState)
+    # We change: a[1] += 123
+    # to:        a[1] = a[1] + 123
+    # We transform the modifier back to its operator
+    # for when users redefine the + function
+
+    operator = let
+        s = string(ex.head)
+        Symbol(s[1:prevind(s, lastindex(s))])
+    end
+    expanded_expr = Expr(:(=), ex.args[1], Expr(:call, operator, ex.args[1], ex.args[2]))
+    return explore!(expanded_expr, scopestate)
+end
+
+function explore_dotprefixed_modifiers!(ex::Expr, scopestate::ScopeState)
+    # We change: a[1] .+= 123
+    # to:        a[1] .= a[1] + 123
+
+    operator = Symbol(string(ex.head)[2:end-1])
+    expanded_expr = Expr(:(.=), ex.args[1], Expr(:call, operator, ex.args[1], ex.args[2]))
+    return explore!(expanded_expr, scopestate)
+end
+
+"Unspecialized mapfoldl."
+function umapfoldl(@nospecialize(f::Function), itr::Vector; init=SymbolsState())
+    if isempty(itr)
+        return init
+    else
+        applied = []
+        # map
+        for e in itr
+            push!(applied, f(e))
         end
+        # reduce
+        out = init
+        for e in applied
+            union!(out, e)
+        end
+        return out
```

Comment:
We can prevent allocating and growing an entire vector here:
```suggestion
        out = init
        for e in itr
            union!(out, f(e))
        end
        return out
```